p2p/src/codec.rs

```diff
@@ -0,0 +1,244 @@
+use crate::core::core::block::{BlockHeader, UntrustedBlockHeader};
+use crate::core::global::header_size_bytes;
+use crate::core::ser::{BufReader, ProtocolVersion, Readable};
+use crate::msg::{Message, MsgHeader, MsgHeaderWrapper, Type};
+use crate::types::{AttachmentMeta, AttachmentUpdate, Error};
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+use core::ser::Reader;
+use std::cmp::min;
+use std::io::Read;
+use std::mem;
+use std::net::TcpStream;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use MsgHeaderWrapper::*;
+use State::*;
+
+const HEADER_IO_TIMEOUT: Duration = Duration::from_millis(2000);
+pub const BODY_IO_TIMEOUT: Duration = Duration::from_millis(60000);
+const HEADER_BATCH_SIZE: usize = 32;
+
+enum State {
+	None,
+	Header(MsgHeaderWrapper),
+	BlockHeaders {
+		bytes_left: usize,
+		items_left: usize,
+		headers: Vec<BlockHeader>,
+	},
+	Attachment(usize, Arc<AttachmentMeta>, Instant),
+}
+
+impl State {
+	fn is_none(&self) -> bool {
+		match self {
+			State::None => true,
+			_ => false,
+		}
+	}
+}
+
+pub struct Codec {
+	pub version: ProtocolVersion,
+	stream: TcpStream,
+	buffer: BytesMut,
+	state: State,
+	bytes_read: usize,
+}
+
+impl Codec {
+	pub fn new(version: ProtocolVersion, stream: TcpStream) -> Self {
+		Self {
+			version,
+			stream,
+			buffer: BytesMut::with_capacity(8 * 1024),
+			state: None,
+			bytes_read: 0,
+		}
+	}
+
+	/// Destroy the codec and return the reader
+	pub fn stream(self) -> TcpStream {
+		self.stream
+	}
+
+	/// Inform codec next `len` bytes are an attachment
+	/// Panics if already reading a body
+	pub fn expect_attachment(&mut self, meta: Arc<AttachmentMeta>) {
+		assert!(self.state.is_none());
+		self.state = Attachment(meta.size, meta, Instant::now());
+	}
+
+	/// Length of the next item we are expecting, could be msg header, body, block header or attachment chunk
+	fn next_len(&self) -> usize {
+		match &self.state {
+			None => MsgHeader::LEN,
+			Header(Known(h)) if h.msg_type == Type::Headers => {
+				// If we are receiving a list of headers, read off the item count first
+				min(h.msg_len as usize, 2)
+			}
+			Header(Known(header)) => header.msg_len as usize,
+			Header(Unknown(len, _)) => *len as usize,
+			BlockHeaders { bytes_left, .. } => {
+				// The header length varies with the number of edge bits. Therefore we overestimate
+				// its size and only actually read the bytes we need
+				min(*bytes_left, header_size_bytes(63))
+			}
+			Attachment(left, _, _) => min(*left, 48_000),
+		}
+	}
+
+	/// Set stream timeout depending on the next expected item
+	fn set_stream_timeout(&self) -> Result<(), Error> {
+		let timeout = match &self.state {
+			None => HEADER_IO_TIMEOUT,
+			_ => BODY_IO_TIMEOUT,
+		};
+		self.stream.set_read_timeout(Some(timeout))?;
+		Ok(())
+	}
+
+	fn read_inner(&mut self) -> Result<Message, Error> {
```

Comment:
Oh interesting. So `read_inner()` will actually return multiple `Message` instances here (each a batch of 32 headers) if we receive is a `Type::Headers`.
Whereas in most cases it only makes sense to call `read_inner()` once to read the full message?

Reply:
Correct!